### DISCUSSION TITLE
Interleave Bytearrays

### DISCUSSION THREAD

**Discussion by A622266** [micropython/micropython#9857] [Hardware & Peripherals] [answered]
I want to interleave two bytearrays.  Since the bytes represent 16-bit numbers, I need to interleave on a 2-byte boundary.  Example
```
bytearray1 = bytearray([1, 1, 2 , 2])
bytearray2 = bytearray([3, 3, 4, 4])

bytearray3 = bytearray(8)
# insert some code to interleave bytearray 1 and bytearray 2 on 2-byte boundary
bytearray3 = bytearray(b'\0x1\0x1\0x3\0x3\0x2\0x2\0x4\0x4)
```
How can this be done in ```Micropython```?

**Comment by dlech** [comment, marked as answer]
One way could be to use a generator function:

```python
def interleave(a, b):
    ia = iter(a)
    ib = iter(b)

    while True:
        try:
            yield next(ia)
            yield next(ia)
            yield next(ib)
            yield next(ib)
        except StopIteration:
            return

bytearray3 = bytearray(interleave([1, 1, 2 , 2], [3, 3, 4, 4]))
```

Or since you are working with 16-bit values, you could use `array` instead if it is available.

```python
from array import array

a1 = array('H', [0x0101, 0x0202])
a2 = array('H', [0x0303, 0x0404])
a3 = array('H')

for n1, n2 in zip(a1, a2):
    a3.append(n1)
    a3.append(n2)
```

**Reply by rkompass** [reply]
Or you use the ptr16 type in viper:

```
@micropython.viper
def interlv_16(first:ptr16, second:ptr16, interlv:ptr16, n:int):
    for i in range(0, n):
        interlv[2*i] = first[i]
        interlv[2*i+1] = second[i]

ba1 = bytearray([1, 1, 2 , 2])
ba2 = bytearray([3, 3, 4, 4])

ba3 = bytearray(8)

interlv_16(ba1, ba2, ba3, 2)

print(ba3)
```

**Reply by A622266** [reply]
thanks dlech and rkompass for these great suggestions.  
@rkompass, can you point me to further reading on the viper decorator?  I haven't seen that before.  I find very little when searching the [https://docs.micropython.org/en/latest/index.html](url)

**Reply by rkompass** [reply]
Did you find the speed section of the [official docs](https://docs.micropython.org/en/latest/reference/speed_python.html?highlight=speed#the-viper-code-emitter) ? There are two nice links in the end (kickstarter).
Basically viper is like C coding in the disguise of python.
You have to consider that viper is in development, and many details are not documented yet. If you use it you may find bugs or inconsistencies, but its worth to point them out. You have to search a lot and sometimes find something.
For example I found [this](https://forum.micropython.org/viewtopic.php?f=6&t=6994&start=10) and [this](https://forum.micropython.org/viewtopic.php?t=727) and there is more in uPys forums or in the issues, [e.g. this](https://github.com/micropython/micropython/issues/8086) if you search there. Then you can try out, like I did in these (poor quality, unedited) examples: [viper.zip](https://github.com/micropython/micropython/files/9946961/viper.zip).
And don't hesitate to formulate your questions here, if you have any (after having searched before, of course). Viper is of interest for many of us. I think it could become a "killer feature" of micropython once it is more complete, because of the enormous speed improvements.

**Reply by rkompass** [reply]
Perhaps I should add that at present viper especially shines at tasks that involve working with the contents of buffers, bytearrays and arrays (information packed continuously in memory). Once you have a pointer and an index viper may access the element and this is done like it is in C. Viper may look up data within more complex python constructs (nested classes or lists for example), but the overhead probably almost is the same as it is in working through the bytecode when the python interpreter is running. So the spectacular > 20x speed improvements cannot be found there.

**Reply by jimmo** [reply]
Yep, this is exactly the sort of thing that `@viper` is for. Thanks @rkompass 

One way to think about how viper works is to think of Python at the lowest level as having a single type -- "object". When you write a+b, even if these are (high level) integers, the compiler still has to say "emit bytecode to load a, load b, call binary-op-add, etc".

The native emitter makes this faster by removing the need to dispatch bytecode, rather it emits machine code that executes those steps. The trick with the viper emitter is that it adds new fundamental integer types in addition to "object" to the compiler so that when it emits a(int) + b(int) it actually just emits the machine code for the add instruction.

(Maybe that was more of an explanation that `@native` isn't like converting your Python code to C, mostly it's just about removing the VM bytecode dispatch, and in fact `@viper` is the thing that lets you write C-in-Python. (Which is what @rkompass said))
